### PR TITLE
Implement transfer operation

### LINKED ISSUE #19: Add signing message support

### PR DESCRIPTION
**Pull request by vcoolish**
Fixes #19

### CODE REVIEW

**Review by MastaBook — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -46,14 +49,62 @@ Override `onActivityResult` to obtain the signing result. Handle the response da
...
+### For all other requests
+
+Add deep link intent filter to your `AndroidManifest.xml`:
+```xml
+<activity
+...
+    <intent-filter android:autoVerify="true">
+    ...
+        <action android:name="android.intent.action.VIEW" />
+        <data android:scheme="trust_sdk" android:host="tx_callback" />
```
Comment (madcake) [inline README.md:61]: Could we use any schema?

File: README.md [hunk 2]
```diff
@@ -46,14 +49,62 @@ Override `onActivityResult` to obtain the signing result. Handle the response da
...
+```
+
+Override 'onNewIntent' if your activity is singleTask or 'onCreate' if not, and handle sdk request callback:
+```kotlin
+override fun onNewIntent(intent: Intent?) {
+    super.onNewIntent(intent)
+    val txCallback = Trust.handleTransactionResult(intent)
```
Comment (madcake) [inline README.md:68]: Could we use Trust.handleTransfer[Swap|Stake]?

File: README.md [hunk 3]
```diff
@@ -46,14 +49,62 @@ Override `onActivityResult` to obtain the signing result. Handle the response da
...
 
+### Sign a transaction
+
+To sign or send a transaction use this code:
+
+```kotlin
+val operation = TransferOperation.Builder().apply {
+      // Required params
+      action = ActionType.SIGN // ActionType.SEND
+      callback = Uri.parse("app_scheme://tx_callback")
```
Comment (madcake) [inline README.md:93]: app_scheme - Show it in Manifest.

File: README.md [hunk 4]
```diff
@@ -46,14 +49,62 @@ Override `onActivityResult` to obtain the signing result. Handle the response da
...
+      assetId = "c60_t0x6B175474E89094C44Da98b954EedeAC495271d0F"
```
Comment (madcake) [inline README.md:94]: Users shouldn't think about asset_id format.

File: README.md [hunk 5]
```diff
@@ -46,14 +49,62 @@ Override `onActivityResult` to obtain the signing result. Handle the response da
...
+      to = "0xF36f148D6FdEaCD6c765F8f59D4074109E311f0c"
+      amount = BigDecimal("1")
```
Comment (madcake) [inline README.md:96]: Add to doc that amount is human readeble.

File: app/src/main/java/com/trustwallet/sdk/MainActivity.kt [hunk 6]
```diff
@@ -19,28 +21,54 @@ class MainActivity : AppCompatActivity() {
...
+            Trust.execute(this, operation)
+        }
+    }
 
-        if (savedInstanceState != null) {
-            getAccountsCall = savedInstanceState.getParcelable("get_accounts_call")
+    override fun onNewIntent(intent: Intent?) {
+        super.onNewIntent(intent)
+        val txCallback = Trust.handleTransactionResult(intent)
+        if (!txCallback?.hash.isNullOrEmpty()) {
```
Comment (madcake) [inline app/src/main/java/com/trustwallet/sdk/MainActivity.kt:57]: When?

File: trust-sdk-client/src/main/java/trust/Trust.kt [hunk 7]
```diff
@@ -21,19 +21,46 @@ object Trust {
...
+        val intent = Intent(Intent.ACTION_VIEW, operation.buildUri())
+        if (canStartActivity(activity, intent)) {
+            activity.startActivity(intent)
+        } else {
+            openMarket(activity)
+        }
+    }
+
+    fun handleTransactionResult(intent: Intent?): TransactionResult? = if (intent?.data?.host == Host.TX_CALLBACK.key) {
+        val signature = intent.data?.getQueryParameter(ExtraKey.TRANSACTION_SIGN.key)
```
Comment (madcake) [inline trust-sdk-client/src/main/java/trust/Trust.kt:39]: User should know what is action and operation. Could we return some object?

File: trust-sdk-client/src/main/java/trust/Trust.kt [hunk 8]
```diff
@@ -21,19 +21,46 @@ object Trust {
...
+        val hash = intent.data?.getQueryParameter(ExtraKey.TRANSACTION_HASH.key)
+        val error = intent.data?.getQueryParameter(ExtraKey.CANCEL.key)
+        when {
+            signature != null -> TransactionResult(signature = signature)
+            hash != null -> TransactionResult(hash = hash)
+            error != null -> TransactionResult(isCancelled = true)
```
Comment (madcake) [inline trust-sdk-client/src/main/java/trust/Trust.kt:45]: Why just canceled?

File: trust-sdk-client/src/main/java/trust/Trust.kt [hunk 9]
```diff
@@ -21,19 +21,46 @@ object Trust {
...
+        null
+    }
+
+    private fun openMarket(activity: Activity) {
+        try {
+            activity.startActivity(Intent(
+                    Intent.ACTION_VIEW,
+                    Uri.parse("market://details?id=${packageName}")))
+        } catch (ignored: ActivityNotFoundException) {
+            activity.startActivity(Intent(
```
Comment (madcake) [inline trust-sdk-client/src/main/java/trust/Trust.kt:58]: Could we open trustwallet.com if market doesn't make available?

File: app/src/main/java/com/trustwallet/sdk/MainActivity.kt (hunk 6 above)
Comment (vcoolish) [reply app/src/main/java/com/trustwallet/sdk/MainActivity.kt:57]: for sign only

File: trust-sdk-client/src/main/java/trust/Trust.kt (hunk 9 above)
Comment (vcoolish) [reply trust-sdk-client/src/main/java/trust/Trust.kt:58]: we use this, because package name is customizable

**Comment by madcake** [comment]
What about tests?

**Comment by vcoolish** [comment]
> What about tests?

updated project, migrated to androidx and added test coverage

**Review by madcake — APPROVED** [review]

**Review by noncash15 — COMMENTED** [review]
app/build.gradle

**Comment by noncash15** [comment]
gh pr checkout 22

**Pull request merged**